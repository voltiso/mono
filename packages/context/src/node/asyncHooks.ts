// â €â“¥ 2024     ðŸŒ©    ðŸŒ©     â €   â €
// â €         ðŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export type AsyncHooksModule = typeof import('async_hooks')
// export type AsyncHooksModule = typeof import('node:async_hooks')

let _asyncHooks: AsyncHooksModule | undefined

export function requireAsyncHooks(): AsyncHooksModule {
	if (!_asyncHooks) {
		try {
			// eslint-disable-next-line n/global-require, unicorn/prefer-module, unicorn/prefer-node-protocol
			const asyncHooks = require('async_hooks') as
				| Partial<AsyncHooksModule>
				| undefined

			// eslint-disable-next-line unicorn/error-message
			if (!asyncHooks?.AsyncLocalStorage) throw new Error()

			_asyncHooks = asyncHooks as never
		} catch {
			throw new Error(
				`@voltiso/context: NodeContext requires 'async_hooks' module, which is not available in this NodeJS version. Please use 'ZoneContext' instead.`,
			)
		}
	}

	return _asyncHooks
}

//

export function haveAsyncHooks(): boolean {
	try {
		requireAsyncHooks()
		return true
	} catch {
		return false
	}
}
