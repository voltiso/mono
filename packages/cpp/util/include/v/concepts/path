#pragma once
#include <v/_/_>

#include <concepts>

namespace VOLTISO_NAMESPACE::concepts {
template <class T>
concept Path = requires(T t) {
	{ t == t } -> std::same_as<bool>;
	{ t / t } -> std::same_as<T>;
};
} // namespace VOLTISO_NAMESPACE::concepts
