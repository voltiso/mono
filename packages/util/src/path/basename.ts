// â €â“¥ 2023     ðŸŒ©    ðŸŒ©     â €   â €
// â €         ðŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €

import type { PathSegmentString, PathString } from '~/object'

export function basename(
	path: PathString<{ separator: '/' }>,
): PathSegmentString<{ separator: '/' }>

export function basename(path: string): PathSegmentString<{ separator: '/' }>

/**
 * Similar to `require('node:path').basename`
 *
 * - Does not detect Windows - only supports UNIX paths
 */
export function basename(path: string) {
	// eslint-disable-next-line no-param-reassign
	while (path.endsWith('/')) path = path.slice(0, -1)
	const lastIdx = path.lastIndexOf('/')
	return path.slice(lastIdx + 1)
}
