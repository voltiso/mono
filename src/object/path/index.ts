export * from './IPath'
export * from './Path'
export * from './get'
export * from './nest'
