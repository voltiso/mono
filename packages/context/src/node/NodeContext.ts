// â €â“¥ 2022     ðŸŒ©    ðŸŒ©     â €   â €
// â €         ðŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €

import { assert, isPromiseLike } from '@voltiso/util'

import { NoContextError } from '~/NoContextError'

import type { Context } from '../Context'
import { requireAsyncHooks } from './asyncHooks'

/**
 * `Context` implementation for NodeJS
 *
 * @internal use `Context` instead
 */
// eslint-disable-next-line etc/underscore-internal
export class NodeContext<T> implements Context<T> {
	readonly _asyncLocalStorage = new (requireAsyncHooks().AsyncLocalStorage)<T>()

	run<Return>(value: T, run: () => Return): Return {
		assert(value !== undefined)

		const oldValue = this._asyncLocalStorage.getStore()

		let result = this._asyncLocalStorage.run(value, run)

		/** Handle async `run` */
		if (isPromiseLike(result)) {
			// if it was `Promise`, not only `PromiseLike`, we would simply use `.finally()`

			// eslint-disable-next-line promise/prefer-await-to-then
			result = result.then(
				value => {
					this._asyncLocalStorage.enterWith(oldValue as never) // T | undefined
					return value
				},
				// eslint-disable-next-line promise/prefer-await-to-callbacks
				error => {
					this._asyncLocalStorage.enterWith(oldValue as never) // T | undefined
					throw error
				},
			) as Return
		}

		return result
	}

	get value(): T {
		const value = this._asyncLocalStorage.getStore()
		if (value === undefined) throw new NoContextError()
		return value
	}
}
