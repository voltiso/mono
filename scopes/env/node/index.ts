import { MyNodeAspect } from './node.aspect';

export type { MyNodeMain } from './node.main.runtime';
export default MyNodeAspect;
export { MyNodeAspect };
