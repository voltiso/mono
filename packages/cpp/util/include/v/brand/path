#pragma once
#include <v/_/_>

#include "v/brand-option"

#include <utility>

namespace VOLTISO_NAMESPACE::brand {
template <bool B = true> struct Path : BrandOption<B> {};
} // namespace VOLTISO_NAMESPACE::brand

// !

namespace VOLTISO_NAMESPACE {
template <class T> constexpr auto is_path = false;

template <class T>
  requires(T::Options::template GET<brand::Path>)
constexpr auto is_path<T> = true;
} // namespace VOLTISO_NAMESPACE

// !

namespace VOLTISO_NAMESPACE::concepts {
template <class T>
concept Path = is_path<T>;
} // namespace VOLTISO_NAMESPACE::concepts

// !

namespace VOLTISO_NAMESPACE {
template <class TPath, class TArg>
  requires(is_path<TPath> && !std::is_const_v<TPath>)
TPath &operator/=(TPath &a, TArg &&b) {
	a <<= "/";
	return a <<= std::forward<TArg>(b);
	// return a <<= '/' <<= std::forward<TArg>(b);
}
} // namespace VOLTISO_NAMESPACE
