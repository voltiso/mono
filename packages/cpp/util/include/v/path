#pragma once
#include <v/_/_>

#include "v/brand/path"
#include "v/dynamic-string"
#include "v/option/num-items"
#include "v/option/self"
#include "v/options"
#include "v/string"
#include "v/tag/explicit-copy-consteval"

#include <cstddef>

// ! the only reason PathString is a separate class is that we can't specify
// deduction guide for alias template, which we need for `operator""_p`
namespace VOLTISO_NAMESPACE {
template <std::size_t NUM_ITEMS>
class Path : public string::Custom<Options<
               option::NUM_ITEMS<NUM_ITEMS>, brand::Path<>,
               option::Self<Path<NUM_ITEMS>>>> {
	using Base = string::Custom<Options<
	  option::NUM_ITEMS<NUM_ITEMS>, brand::Path<>,
	  option::Self<Path<NUM_ITEMS>>>>;
	using Base::Base;

public:
	// `consteval`-only constructor for `operator""_p`
	template <std::size_t NUM_ITEMS_WITH_NULL>
	  requires(NUM_ITEMS_WITH_NULL == NUM_ITEMS + 1)
	consteval Path(const char (&items)[NUM_ITEMS_WITH_NULL])
	    : Base(tag::EXPLICIT_COPY_CONSTEVAL, items) {
		VOLTISO_EQ(items[NUM_ITEMS], '\0');
	}
};

// deduction guide
template <std::size_t NUM_CHARS>
Path(const char (&)[NUM_CHARS])
  -> Path<NUM_CHARS - 1>; // -1 for null terminator
} // namespace VOLTISO_NAMESPACE

// !

namespace VOLTISO_NAMESPACE::path {
using Dynamic = dynamicString::Custom<Options<brand::Path<>>>;
} // namespace VOLTISO_NAMESPACE::path

// !

// namespace VOLTISO_NAMESPACE {
// template <std::size_t NUM_CHARS>
// using PathString =
//   string::Custom<Options<option::NUM_CHARS<NUM_CHARS>, brand::Path>>;

// // deduction guide
// template <size_t N> PathString(const char (&)[N]) -> PathString<N - 1>;

// } // namespace VOLTISO_NAMESPACE

// !

namespace VOLTISO_NAMESPACE {
// #pragma GCC diagnostic push
// #pragma GCC diagnostic ignored "-Wuser-defined-literals"
template <Path PATH> consteval auto &operator""_p() { return PATH; }
// #pragma GCC diagnostic pop
} // namespace VOLTISO_NAMESPACE
