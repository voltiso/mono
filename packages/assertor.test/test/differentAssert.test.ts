// â €â“¥ 2024     ðŸŒ©    ðŸŒ©     â €   â €
// â €         ðŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €

import * as fs from 'node:fs/promises'
import * as path from 'node:path'

describe('differentAssert', () => {
	it('works', async () => {
		expect.hasAssertions()

		const file = await fs.readFile(
			// eslint-disable-next-line unicorn/prefer-module
			path.join(__dirname, '../dist/esm/differentAssert.js'),
		)

		expect(file.toString()).toMatchSnapshot()
	})
})
