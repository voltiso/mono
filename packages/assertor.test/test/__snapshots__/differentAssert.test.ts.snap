// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`differentAssert works 1`] = `
"// â €â“¥ 2025     ğŸŒ©    ğŸŒ©     â €   â €
// â €         ğŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €
import assert from 'node:assert';
export function differentAssert() {
    assert(1);
}
//# sourceMappingURL=differentAssert.js.map"
`;
